dataplane/api-server/src/config.rs

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+*/
+use anyhow::{anyhow, Result};
+use clap::Parser;
+use std::path::PathBuf;
+
+#[derive(Debug, Parser, Clone)]
+pub struct TLSConfig {
+    #[clap(long, default_value = "false")]
+    pub enable_tls: bool,
+    #[clap(long, default_value = "false")]
+    pub enable_mtls: bool,
+    pub server_certificate_path: Option<PathBuf>,
+    pub server_private_key_path: Option<PathBuf>,
+    pub client_certificate_authority_root_path: Option<PathBuf>,
```

Comment:
So in general I've seen a lot of `Option<T>` fields in struct attributes over the years and that's fine in general. What's interesting in this case is that you're not trying to express that these fields are ultimately optional, but instead that they are optional situationally based on the values of other fields. In effect, this means that you have a couple of different _variants_ of a `TLSConfig` that are acceptable.

Given that notion, and the fact that I immediate thought "variants", I think you should consider whether an [Enumeration](https://doc.rust-lang.org/rust-by-example/custom_types/enum.html) might actually be a better fit here, where we would have a `TLSConfig::ServerOnlyTLS` variant and a `TLSConfig::MutualTLS`.

Now... the only problem with this is that I think the standards really don't accept `ServerOnlyTLS` as official nomenclature everywhere they often just say `TLS`, so... that might require some finesse but either way just think this over and let me know what you think.

Reply:
Why not just `TLSConfig::TLS` and `TLSConfig::MutualTLS`? Those correspond to the `TLS` and `mTLS` acronyms, respectively, which are standard afaik.

Reply:
I'm good with those names :+1: